dd: basic implementation

bmusin opened the pull request:
Hi. It's initial implementation, Converting blocks and handling other options on the way.

Currently, it has problems with signal handling, doesn't resume copying if interrupted.

Arcterus commented on src/dd/util.rs:
```diff
@@ -0,0 +1,180 @@
+        &mut sin
+    } else {
+        let inp = File::open(&i_name);
+        if inp.is_err() {
```
This would probably be better as a `match`.  Same for the one down below as well.

bltnkl replied:
Ok.

Arcterus commented on src/dd/util.rs:
```diff
@@ -0,0 +1,180 @@
+            break;
+        }
+    }
+    display_progress(Arc::clone(&cf));
```
I don't think you need to clone here.

bltnkl replied:
IMO, It needs to, because display_progress is also used by signal handler, which also gets data in shared pointer.

Arcterus replied:
But you're not doing anything with this `Arc` in this function anymore, so can't you just pass ownership to `display_progress()`?

bltnkl replied:
display_progress is also called by "main" thread. I made that signature, so that there would be unified interface to both main and called threads.

Arcterus replied:
Does the compiler throw an error if you remove the clone?

Arcterus commented on src/dd/util.rs:
```diff
@@ -0,0 +1,180 @@
+    copy(Arc::clone(&cf));
+}
+
+pub fn err(s: String) {
```
There's actually a macro for this in `uucore` (we should probably document this somewhere).  You can use `crash!()` or `crash_if_err!()` to do the same thing as this function, but with a given exit code and a standard error prefix (something like `dd: error:` in this case).

bltnkl replied:
Document it, please, many will benefit.

Arcterus replied:
👍 

I'll try to do so soon.

bltnkl commented on src/dd/args.rs:
```diff
@@ -0,0 +1,412 @@
+use super::util::{block, def, err, unblock,};
+use super::defs::{Arg, Config};
+
+pub fn parse_args() -> Arc<Mutex<Config>> {
```
I'll try, as soon as I get program up to some valuable state,
arguments are of type key=value without dashes in front, so I font,
tools like clap and getopt(s) aren't very relevant here.

Actually, only top paragraph is parsing, the lower parts are about "sanity checking"(OpenBSD).

bltnkl commented on src/dd/args.rs:
```diff
@@ -0,0 +1,412 @@
+use super::defs::{Arg, Config};
+
+pub fn parse_args() -> Arc<Mutex<Config>> {
+    let mut cf = Box::new(Config::default());
```
I can't use mere stack variable because it needs to be modified both in parse_args and handlers.
So I use box to pass ownership to handler and back via return value from handler.

Yes, it's cumbersome, but otherwise one should resort to unsafe code as it's not possible to get two mutable links to one object.

Arcterus replied:
Why can't you pass it as a mutable reference to the handlers?

bltnkl commented on src/dd/args.rs:
```diff
@@ -0,0 +1,412 @@
+        let pair: Vec<&str> = arg.splitn(2, '=').collect();
+        if pair.len() == 2 {
+            let (key, val) = (pair[0], pair[1]);
+            let arg = match ARGS.binary_search_by(|probe| probe.name.cmp(&key)) {
```
I blindly copied it from OpenBSD's dd. I'll try to refactor it.

bltnkl commented on src/dd/args.rs:
```diff
@@ -0,0 +1,412 @@
+    }
+
+    let mut pref = val;
+    let mut size: usize = match val.chars().nth(val.len() - 1) {
```
Ok.

bltnkl commented on src/dd/dd.rs:
```diff
@@ -0,0 +1,38 @@
+use util::{display_progress, run};
+
+pub fn uumain(_args: Vec<String>) -> i32 {
+    let signal = chan_signal::notify(&[Signal::INT]);
```
What crate should I use for signal handling? Or, maybe, uutils' one?

Arcterus replied:
It’s up to you.  The signals stuff in `uucore` isn’t really for handling signals.  It’s only really used by `kill` to send signals IIRC.

bltnkl commented on src/dd/util.rs:
```diff
@@ -0,0 +1,180 @@
+
+use super::defs::{Config};
+
+fn dd_out(buf: &mut Vec<u8>, pos: usize, num: usize, out: &mut ::std::io::Write) -> Option<usize> {
```
Ok.

Arcterus commented on src/dd/util.rs:
```diff
@@ -0,0 +1,180 @@
+}
+
+pub fn run(cf: Arc<Mutex<Config>>, _sdone: chan::Sender<()>) {
+    copy(Arc::clone(&cf));
```
I think this is an unnecessary clone was well.

bltnkl commented:
It handles if,of,bs,ibs,obs,count arguments. Skip,seek partially. Doesn't work for stdin yet.

bltnkl commented:
Fixed skip interpretation.

Arcterus commented on src/dd/dd.rs:
```diff
@@ -0,0 +1,640 @@
+                    Some(m) => {
+                        cf.skip = m;
+                    }
+                    _ => err(format!("invalid {}: {}", key, val)),
```
You might want to make a function or something to do this sort of stuff.  Most of the `match`es seem pretty similar.

bmusin replied:
Did.

Arcterus commented on src/dd/dd.rs:
```diff
@@ -0,0 +1,640 @@
+fn setup(cf: &mut Conf) {
+    let inp = if cf.is_stdin {
+        Some(InpFileWrap {
+            f: InpFileType::Stdin(io::stdin()), // TODO: use .lock() if necessary
```
You should indeed probably use `.lock()` so you only have to lock `stdin`'s mutex once.  Same for `stdout`.

Arcterus commented on src/dd/dd.rs:
```diff
@@ -0,0 +1,640 @@
+        }
+    }
+
+    let mut buf = vec![0u8; cmp::min(cf.inp_block_size, cf.out_block_size)];
```
I would move the stuff from here onward into another function.  IMO in `uumain()` you should basically do something like:
```rust
let (input, output) = setup(&mut conf);
copy(input, output, &conf);
```

bmusin replied:
Did.

Arcterus commented on src/dd/dd.rs:
```diff
@@ -0,0 +1,640 @@
+    Some(size)
+}
+
+pub fn get_i(val: &str) -> Option<isize> {
```
This function is very similar to `get_u()`, so you should probably factor out the common code into a function or trait.  Also, I'm not 100% sure, but there may be crates on [crates.io](https://crates.io) that can do this for you.

Arcterus replied:
Yeah as far as I can tell there is only one line that is meaningfully different (the first `if` expression).

Arcterus commented on src/dd/dd.rs:
```diff
@@ -0,0 +1,640 @@
+        cf.inp_full, cf.inp_part, cf.out_full, cf.out_part
+    );
+
+    match cf.trunc {
```
In this case, I think it makes more sense to just do:
```rust
if cf.trunc != 0 {
    eprintln!(
        "{} truncated record{}",
        cf.trunc,
        if cf.trunc == 1 { "" } else { "s" }
    );
}
```

bmusin replied:
Did.

Arcterus commented on src/dd/dd.rs:
```diff
@@ -0,0 +1,640 @@
+        let mut pos = 0;
+        let mut left = cf.inp_block_size;
+        let mut session = 0;
+        'read_block: loop {
```
If you never `break` the outer loop in this inner loop, you might want to just factor the inner loop out into a separate function.

bmusin replied:
I tried, unfortunately it introduces even more complications.

Arcterus commented on src/dd/dd.rs:
```diff
@@ -0,0 +1,640 @@
+                break;
+            }
+            if let Ok(m) = out.write(&buf[pos..pos + num]) {
+                {
```
Is this block necessary?

bmusin replied:
Removed.

Arcterus commented on src/dd/dd.rs:
```diff
@@ -0,0 +1,640 @@
+                Ok(0) => is_read_0 = true,
+                Ok(num) => {
+                    session += num;
+                    {
```
Is this block necessary?

bmusin replied:
Removed.

Arcterus commented on src/dd/dd.rs:
```diff
@@ -0,0 +1,745 @@
+impl Conf {
+    const DEFAULT_BLOCK_SIZE: usize = 512;
+
+    pub fn set_if(&mut self, if_: &str) {
```
I would prefer this just be `iflag` or something instead of `if_`.  Same for the `of_` below.

bmusin replied:
Did.

Arcterus commented on src/dd/dd.rs:
```diff
@@ -0,0 +1,745 @@
+            match key {
+                "if" => cf.set_if(val),
+                "of" => cf.set_of(val),
+                "ibs" => if let Some(m) = get_u(val) {
```
You might want to just create another `match` for all of these options that basically have the same code (to reduce the boilerplate).  So maybe do something like:
```rust
match key {
    "if" => cf.set_if(val),
    "of" => cf.set_of(val),
    "conv" => {
        for elem in val.split(",") {
            match Conf::get_flag(elem) {
                Conf::INVALID_FLAG => err("invalid flag passed to conv"),
                flag => cf.set_flag(flag),
            }
        }
    }
    _ => {
        if let Some(m) = get_u(val) {
            match key {
                "ibs" => cf.inp_block_size = m,
                "obs" => cf.out_block_size = m,
                "bs" => {
                    cf.inp_block_size = m;
                    cf.out_block_size = m;
                }
                "cbs" => cf.cnv_block_size = m,
                "skip" => cf.skip = m,
                "seek" => cf.seek = m,
                "count" => cf.count = m
                _ => err(format!("invalid argument: {}", key))
            }
        } else {
            err(format!("invalid {}: {}", key, val));
        }
    }
}
```

Although doing something like the above would mostly likely slow down the code as you are now matching twice rather than once.

bmusin replied:
Speed doesn't matter, as this small amount of arguments is parsed only once, maintanability is a way more important. Thanks
for useful tip.

bmusin replied:
It's logical error here: value is checked before checking that key exists.

Arcterus replied:
I don't think it matters either way as the GNU version gives the same error in both cases (it's like "invalid operand: key=value" or something).

Arcterus commented on src/dd/dd.rs:
```diff
@@ -0,0 +1,745 @@
+        }
+    }
+
+    // Behaviour is unspecified. We choose to terminate.
```
Does the GNU version terminate as well?

bmusin replied:
I haven't read yet.

Arcterus replied:
Just an FYI but don't read the GNU source as it's licensed under the GPL.

Arcterus commented on src/dd/dd.rs:
```diff
@@ -0,0 +1,745 @@
+    let cf2 = Arc::clone(&cf1);
+
+    let handle = thread::spawn(move || {
+        let cf3 = Arc::clone(&cf1);
```
This clone isn't necessary.

bmusin replied:
Sadly, I don't know how to do it another way. I clone because it's also used by signal handler,
so I wrap value in smart pointer and pass around copies of that Arc.

Arcterus replied:
You aren't using `cf1` anywhere else in the function though, so AFAICT you should be able to just use `cf1` directly.

bmusin replied:
Fixed.

Arcterus commented on src/dd/dd.rs:
```diff
@@ -0,0 +1,745 @@
+    let handle = thread::spawn(move || {
+        let cf3 = Arc::clone(&cf1);
+        let (mut inp, mut out) = setup(Arc::clone(&cf3));
+        copy(inp, out, Arc::clone(&cf3), s_done);
```
This clone isn't necessary.

bmusin replied:
Sadly, I don't know how to do it another way. I clone because it's also used by signal handler,
so I wrap value in smart pointer and pass around copies of that Arc.

Arcterus replied:
Same as above.  You've already cloned `cf3` for `setup()`, so you should be able to pass `cf3` to `copy()` directly.  Is the signal handler doing something strange that's causing issues?

bmusin replied:
Fixed. cf1 will be send to thread, cf2 will be used by signal handler.

Arcterus commented on src/dd/dd.rs:
```diff
@@ -0,0 +1,745 @@
+
+    chan_select! {
+        signal.recv() -> signal => {
+            display_stat(Arc::clone(&cf2));
```
This clone isn't necessary.

bmusin replied:
Sadly, I don't know how to do it another way. I clone because it's also used by signal handler,
so I wrap value in smart pointer and pass around copies of that Arc.

Arcterus commented on src/dd/dd.rs:
```diff
@@ -0,0 +1,745 @@
+        }
+    }
+    drop(cf);
+    display_stat(Arc::clone(&first));
```
This clone isn't necessary.

bmusin replied:
Sadly, I don't know how to do it another way. I clone because its also used signal handler,
so I don't pass it by value.

bmusin replied:
Maybe  you will share an idea?

Arcterus replied:
You cloned it earlier so you should be able to just pass ownership of the `Arc` to `display_stat()` (unless this causes a compiler error for you?  If it does let me know so I can look more in-depth).

Arcterus commented on src/dd/dd.rs:
```diff
@@ -0,0 +1,745 @@
+
+    flags: u32,
+
+    short_2_short: bool,
```
I think I would prefer if `2` was `to`, but I'll leave it up to you.

bmusin replied:
Did.

bmusin commented:
Removed clone operations.

Arcterus commented on src/dd/dd.rs:
```diff
@@ -0,0 +1,734 @@
+    }
+
+    pub fn is_flag_valid(flag: &str) -> bool {
+        flag == "ibs" || flag == "obs" || flag == "bs" || flag == "cbs" || flag == "if"
```
This is a situation where I might use a `match` to return a `bool`.

bmusin replied:
Fixed.

Arcterus commented on src/dd/dd.rs:
```diff
@@ -0,0 +1,734 @@
+
+    if !(cf.flags & Conf::BLOCK != 0 || cf.flags & Conf::UNBLOCK != 0 || cf.flags & Conf::LCASE != 0
+        || cf.flags & Conf::UCASE != 0 || cf.flags & Conf::SWAB != 0)
+        {
```
The braces should line up with the `if`.  Also the second line of conditions should probably be indented one more time.

bmusin replied:
@Arcterus, _rustfmt_ leaves indentation of second line as is.
Should I indent it one more time?

Arcterus replied:
It’s fine as is.

bmusin replied:
Nice, I left it indented once.

Arcterus commented on src/dd/dd.rs:
```diff
@@ -0,0 +1,734 @@
+    };
+    let mut out = out.unwrap();
+
+    match inp.f {
```
Can you move this and the `match out.f` into separate functions?

bmusin replied:
Moved.

Arcterus commented on src/dd/dd.rs:
```diff
@@ -0,0 +1,749 @@
+    _sdone: chan::Sender<()>,
+) -> result::Result<(), String> {
+    {
+        let mut cf = cf_ptr.lock().unwrap();
```
In order to prevent the signal handling thread from blocking until the end of the program, you may need to reacquire the lock on every loop iteration.  If you do so it might be better to just avoid using `chan_select!` and spawn a channel using `chan::async()`, which you then pass to `notify_on()` for each of the signals.  You would then have to check at the bottom of the `for` loop here if a signal was passed and react accordingly.

bmusin commented on src/dd/dd.rs:
```diff
@@ -0,0 +1,749 @@
+            if out.is_seekable {
+                if cf_guard.seek == 0 {
+                    if !cf_guard.has_flag(Conf::NOTRUNC) {
+                        match f.seek(io::SeekFrom::Start(cf_guard.seek as u64)) {
```
Changed.

bmusin commented on src/dd/dd.rs:
```diff
@@ -0,0 +1,749 @@
+fn setup(cf: Arc<Mutex<Conf>>) -> (InpFileWrap, OutFileWrap) {
+    let mut cf_guard = cf.lock().unwrap();
+
+    let inp = if cf_guard.is_stdin {
```
Moved.

bmusin commented on src/dd/dd.rs:
```diff
@@ -0,0 +1,749 @@
+            is_seekable: false,
+        })
+    } else {
+        let f = File::open(cf_guard.inp_name.as_ref() as &str);
```
Replaced.

bmusin commented on src/dd/dd.rs:
```diff
@@ -0,0 +1,749 @@
+
+    drop(cf_guard);
+    inp = seek_inp(inp, Arc::clone(&cf));
+    out = seek_out(out, Arc::clone(&cf));
```
Got rid of it.

bmusin commented on src/dd/dd.rs:
```diff
@@ -0,0 +1,749 @@
+            }
+        }
+    }
+    display_stat(cf_ptr);
```
I created nested {} scope for it, it'is was being dropped automatically.
If  I remove nested scope at do drop before *display_stat* it errs, but works if NLL is enabled.
Should I use NLL feature or leave nested {} scope as is?

Arcterus replied:
I didn't see the nested scope when I made that comment, so keep it as is (NLL only works on nightly right now).

bmusin replied:
Did.

bmusin commented on src/dd/dd.rs:
```diff
@@ -0,0 +1,749 @@
+
+    let handle = thread::spawn(move || {
+        let (mut inp, mut out) = setup(Arc::clone(&cf1));
+        copy(inp, out, Arc::clone(&cf1), s_done);
```
Got rid of.

bmusin commented on src/dd/dd.rs:
```diff
@@ -0,0 +1,749 @@
+    {
+        return None;
+    }
+    for m in pref.split('x').collect::<Vec<&str>>() {
```
Changed.

bmusin commented on src/dd/dd.rs:
```diff
@@ -0,0 +1,749 @@
+    } else {
+        let f = File::open(cf_guard.inp_name.as_ref() as &str);
+        if f.is_err() {
+            err("cannot open input file".to_string());
```
Fixed.

bmusin commented on src/dd/dd.rs:
```diff
@@ -0,0 +1,749 @@
+            display_stat(cf2);
+            exit(1);
+        },
+        r_done.recv() => ()
```
Should I continue copying even after receiving signal?
Standard says: "For SIGINT, the dd utility shall interrupt its current processing, write status information to standard error, and exit as though terminated by SIGINT"

Arcterus commented on src/dd/dd.rs:
```diff
@@ -0,0 +1,679 @@
+    Some(size)
+}
+
+pub fn err<S: Into<String>>(s: S) {
```
You should probably change this to:
```rust
pub fn err<S: Into<String>>(s: S) -> ! {
    // stuff
}
```

This will let you avoid adding `None` and stuff in cases like:
```rust
if x {
    Some(0)
} else {
    err("err");
    None
}
```

bmusin replied:
Changed.

Arcterus commented on src/dd/dd.rs:
```diff
@@ -0,0 +1,679 @@
+
+pub fn parse_args() -> Conf {
+    let mut cf = Conf::default();
+    for arg in env::args().skip(1) {
```
This function should be like:
```rust
pub fn parse_args(args: Vec<String>) -> Conf {
    let mut cf = Conf::default();
    for arg in args.into_iter().skip(1) {
        // stuff
    }
    // stuff
}
```

You shouldn't use `env::args()` because the args given to `uumain()` have the first element removed if the executable was started using the multicall binary (the `uutils` binary).  `env::args()` will return the list of args with `uutils` still present.

bmusin replied:
Changed.

Arcterus commented on src/dd/dd.rs:
```diff
@@ -0,0 +1,679 @@
+    }
+
+    // Behaviour is unspecified. We choose to terminate.
+    if cf.count == 0 {
```
This causes the program to fail if you run it like `dd if=infile of=outfile`.  You probably want to wrap this in an `Option` with `None` being the default.  It should only fail if `cf.count == Some(0)`.

bmusin replied:
Changed.

Arcterus commented on src/dd/dd.rs:
```diff
@@ -0,0 +1,679 @@
+        let mut cf = cf_ptr.lock().unwrap();
+        let mut buf = vec![0u8; cmp::min(cf.inp_block_size, cf.out_block_size)];
+        let mut total = 0;
+        for _ in 0..cf.count {
```
This should behave like an infinite loop which breaks when everything has been copied if `cf.count == None`.

Arcterus replied:
Also to avoid the display status thread from blocking until the end, you are definitely going to need to reacquire the mutex sometimes (either every loop iteration or every few seconds or something like that).  You could also consider using a `RwLock` as the main thread is the only one that modifies the struct (and it only gets modified in a couple spots in this function).

bmusin replied:
Changed.

bmusin commented:
@Arcterus, could you please describe preciously, how program should handle and what signals?

Arcterus commented:
Do you mean what signals should be handled and how they should be handled?

bmusin commented:
@Arcterus, yes.

Arcterus commented:
`SIGINFO` on BSD and macOS, `SIGUSR1` on Linux: print current status/progress info to `stderr`.
`SIGINT`: print status/progress info and exit

That's all the signals that need to be handled (well, as far as I know at the moment).

bmusin commented:
@Arcterus, chan_signal doesn't support SIGINFO.

Arcterus commented:
Just write a comment saying so for now.  It can be added later.

bmusin commented:
@Arcterus, I put comment.

Arcterus commented:
Sorry it's taken so long to get back to you, but can you write some tests comparing the behavior to GNU dd?  I am busy with upcoming midterms and essays so I don't have time to compare at the moment.

bmusin commented:
@Arcterus, what do you exactly mean? Should I add integration tests that run this program on a set of predefined files and compare that everything is correct?

Arcterus commented:
Yeah, that's pretty much what I mean.  Just make sure the behavior is the same as GNU dd.

bmusin commented:
@Arcterus, could you please clarify by providing examples? I think that GNU's dd's behaviour is pretty standard.

Arcterus commented:
I’m just saying if there are differences between implementations (_e.g._ GNU and OpenBSD) do whatever the GNU version does.

bmusin commented:
@Arcterus, it's no longer OpenBSD port anymore.
1. Changed default block size to 1024, like GNU's dd does.
2. On BSD's in has stub for handling SIGINFO, which isn't supported by chan_signal yet.
\- was already done, changed order of commands.
3. On SIGINT prints progress and exits - was already done.

Arcterus commented:
Can you still add some basic tests?  Like, for example, `dd if=testfile of=otherfile`.  It would be good to do so to avoid potential regressions and so on.  Also, why are `src/join/join.rs` (and other files) marked as changed?

mmstick commented:
Is this implementation going to make the same major flaw that GNU's `dd` does? Basically, it has an issue where dd destroys your data if the offset from the start sector overlaps with the end sector. The solution is to write sectors in reverse in those situations, to guarantee that copies aren't overlapping.

Essentially, it will start at the beginning, read N amount of sectors, and then seek to a point before the end sector and obliterate whatever data was written there. Eventually it catches up and starts reading the already-written data and copying that until to end is reached. Any data in the crossfire is lost. Using a simple check, you can avoid that so that you start reading from the end sector, then write to the new end sector, in reverse. No more overlapping reads & writes.

Arcterus commented:
@mmstick this is the sort of issue where I would consider differing in behavior from GNU (or at least provide the ability to change the behavior at compile-time/runtime if we don't differ by default).  In any case, given that this PR hasn't been updated in a couple months, I will likely make some changes to the code and then merge an implementation of `dd` in the next few weeks.  I will take this issue into consideration when I do that.

Luke-Nukem commented:
It's been over a year for this - is help required?

sylvestre commented:
Is there still an interest to commit this?

sylvestre commented:
closing as it is stuck!
Feel free to reopen
